install_tools.sh

```diff
@@ -55,25 +55,31 @@ clone() {
     echo "Cloned $TOOLS_REPO@$BRANCH into local repo: $TOOLS_PATH"
 }
 
-add_to_path() {
-    if [ "$#" -lt 1 ]; then
-        echo "Error: You need to provide a path."
-        echo "Usage: ${FUNCNAME[0]} <path>"
-        return 1
-    fi
-    local PATH_TO_ADD=$(readlink -m $1)
-    if [[ ":$PATH:" != *":$PATH_TO_ADD:"* ]]; then
-        export PATH="$PATH:$PATH_TO_ADD"
-        echo "Added $PATH_TO_ADD to PATH"
-    fi
-}
-
 parse_args $@
 fetch || (rm -rf $TOOLS_PATH && clone)
 
-# install scriptlets (i.e. add them to path)
-add_to_path $TOOLS_PATH/scriptlets
+# add scriptlets to agent's PATH
+SCRIPTLETS_PATH=$TOOLS_PATH/scriptlets
+source "$SCRIPTLETS_PATH/defs/add_to_path"
+add_to_path $SCRIPTLETS_PATH
+add_to_path $SCRIPTLETS_PATH/sru-helpers
 
-# install launcher
-pip -q install $TOOLS_PATH/cert-tools/launcher
+# figure out where to place scriptlets on the device
+REMOTE_PATH=$(cat $SCRIPTLETS_PATH/defs/scriptlet_path | _run bash)
+[ $? -eq 0 ] || exit 1
+
+# copy selected scriptlets over to the device
+# and then move them somewhere in the device's PATH
+DEVICE_SCRIPTLETS=(retry check_for_packages_complete wait_for_packages_complete install_packages)
+_put "${DEVICE_SCRIPTLETS[@]/#/$SCRIPTLETS_PATH/}" --
+_run sudo mv "${DEVICE_SCRIPTLETS[@]}" $REMOTE_PATH
+
+# fuser is required by `check_for_packages_complete`
+# (so install it on both the agent and the device)
+$SCRIPTLETS_PATH/defs/install_psmisc
+cat $SCRIPTLETS_PATH/defs/install_psmisc | _run bash
+[ $? -eq 0 ] || exit 1
+
+# install launcher tool on the agent
 add_to_path ~/.local/bin
+pip -q install $TOOLS_PATH/cert-tools/launcher
```

Comment:
Does this get installed in a virtualenv? it might accidentally work for now because the agent host container we use is based on focal I think, but if this ever gets updated to more recent versions, then it will fail.

Reply:
No, I can change it so that it does. But can you just explain why it will fail on more recent versions?

Reply:
There's a great write-up about it [here](https://www.omgubuntu.co.uk/2023/04/pip-install-error-externally-managed-environment-fix). Anything >= 23.10 will break with just doing "pip install" I think

Reply:
Thanks Paul, that's very useful. I will address this.

Reply:
@plars I have confirmed this will be a problem in future Ubuntu versions and researched possible solutions. My suggestion would be to use `pipx`, as long as we have it installed previously. The change to the existing code is minimal and it seems to fit our use-case perfectly ("install packages with cli entrypoints"). How would you feel about that?

Reply:
+1